clis/teliod/src/main.rs

```diff
@@ -115,17 +152,188 @@ impl CommandListener {
     }
 }
 
+async fn init_api(
+    config: &mut TeliodDaemonConfig,
+    config_path: String,
+) -> Result<ClientConfig, TeliodError> {
+    let mut client_config = ClientConfig::new(config).await?;
+
+    if client_config.machine_identifier.eq("") {
+        client_config.machine_identifier =
+            match load_identifier_from_api(&config.authentication_token, client_config.public_key)
+                .await
+            {
+                Ok(id) => id,
+                Err(e) => {
+                    debug!("Unable to load identifier due to {e}. Registering ...");
+                    register_machine(
+                        &client_config.hw_identifier,
+                        client_config.public_key,
+                        &client_config.auth_token,
+                    )
+                    .await?
+                }
+            }
+    }
+
+    let mut retries = 0;
+    loop {
+        let status = update_machine(&client_config).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(TeliodError::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                debug!("Unable to update. Registering machine ...");
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                client_config.machine_identifier = register_machine(
+                    &client_config.hw_identifier,
+                    client_config.public_key,
+                    &client_config.auth_token,
+                )
+                .await?;
+            }
+            _ => {
+                // Retry with exp back-off
+                if retries < MAX_RETRIES {
+                    let backoff_time = BASE_DELAY_MS * 2_u64.pow(retries);
+                    let sleep_time = min(backoff_time, MAX_BACKOFF_TIME);
+                    debug!("Retrying after {} ms...", sleep_time);
+                    tokio::time::sleep(Duration::from_millis(sleep_time)).await;
+                    retries += 1;
+                } else {
+                    // If max retries exceeded, exit daemon
+                    return Err(TeliodError::UpdateMachineTimeout);
+                }
+            }
+        }
+    }
+
+    config.private_key = Some(client_config.private_key.clone());
+    config.hw_identifier = Some(client_config.hw_identifier.clone());
+    config.machine_identifier = Some(client_config.machine_identifier.clone());
+
+    let mut file = File::create(config_path)?;
+    serde_json::to_writer(&mut file, &config)?;
```

Comment:
It's rather strange for a program to modify it's own config - I think you should either move it to another file, or just don't save it at all (@Jauler has some smarter ideas how to do it, maybe you can create some task for researching the topic)

Reply:
Moved to its own file in local dir